### PR TITLE
Use Windows native API to impl dispatcher

### PR DESCRIPTION
**Pull request by JunkuiZhang**
Using `Threadpool` and `TimerQueue` which are provided by the native Windows APIs, to implement the corresponding interfaces, we do not need to sort tasks ourselves as Windows will handle it in a relatively more efficient manner, I guess. I am unsure if Zed would welcome this PR, and suggestions are welcome.

Release Notes:

- N/A

### CODE REVIEW

File: crates/gpui/src/platform/windows/dispatcher.rs [hunk 1]
```diff
@@ -133,27 +122,64 @@ impl PlatformDispatcher for WindowsDispatcher {
...
-    instant: Instant,
+fn init() -> PTP_POOL {
+    unsafe {
+        let threadpool = CreateThreadpool(None);
+        if threadpool.0 == 0 {
+            log::error!(
+                "unable to initialize a thread pool: {}",
+                std::io::Error::last_os_error()
+            );
+            panic!();
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/dispatcher.rs:133]: Let's get rid of this `log::error` and move the message into `panic!()`

File: crates/gpui/src/platform/windows/dispatcher.rs [hunk 2]
```diff
@@ -133,27 +122,64 @@ impl PlatformDispatcher for WindowsDispatcher {
     }
 }
 
-struct RunnableAfter {
-    runnable: Runnable,
-    instant: Instant,
+fn init() -> PTP_POOL {
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/dispatcher.rs:125]: Can we just inline this whole function into `WindowsDispatcher::new()`?

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
Glad you're integrating this :D

**Comment by JunkuiZhang** [comment]
> Glad you're integrating this :D

Really happy to hear this feeadback!

**Review by mikayla-maki — APPROVED** [review]

File: crates/gpui/src/platform/windows/dispatcher.rs [hunk 3]
```diff
@@ -99,25 +81,45 @@ impl PlatformDispatcher for WindowsDispatcher {
     }
 
     fn dispatch(&self, runnable: Runnable, label: Option<TaskLabel>) {
-        self.background_sender
-            .send((runnable, label))
-            .inspect_err(|e| log::error!("Dispatch failed: {e}"))
-            .ok();
+        let _ = self.dispatch_on_threadpool(runnable);
```
Comment (zaucy) [inline crates/gpui/src/platform/windows/dispatcher.rs:84]: Error probably shouldn't be ignored. Maybe log any of these `dispatch_on_threadpool()` calls?

```suggestion
        self.dispatch_on_threadpool(runnable).log_err();
```

**Review by zaucy — CHANGES_REQUESTED** [review]
I love the idea of using the win32 thread pool for the background tasks :+1:

File: crates/gpui/src/platform/windows/dispatcher.rs (hunk 3 above)
Comment (zaucy) [reply crates/gpui/src/platform/windows/dispatcher.rs:84]: what I mean is you're ignoring the `Result` of `dispatch_on_threadpool`. I'm not against you use `last_os_error()`, but I'd recommend not returning a `Result` if it's just going to be ignored anyways.

Comment (JunkuiZhang) [reply crates/gpui/src/platform/windows/dispatcher.rs:84]: > what I mean is you're ignoring the `Result` of `dispatch_on_threadpool`. I'm not against you use `last_os_error()`, but I'd recommend not returning a `Result` if it's just going to be ignored anyways. Preferrably would be to keep the `Result` and simply handle (just log it) at the call site.

I got your idea, thanks!

**Pull request merged**